src/System.Net.Security/src/System/Net/Security/SniHelper.cs

```diff
@@ -0,0 +1,278 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Globalization;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace System.Net.Security
+{
+    internal class SniHelper
+    {
+        private static IdnMapping s_idnMapping = CreateIdnMapping();
+
+        public static string GetServerName(byte[] clientHello)
+        {
+            return GetSniFromSslPlainText(clientHello);
+        }
+
+        // https://tools.ietf.org/html/rfc6101#section-5.2.1
+        // SSLPlainText structure:
+        //   - ContentType (1 byte) => 0x16 is handshake
+        //   - ProtocolVersion version (2 bytes)
+        //   - uint16 length
+        //   - opaque fragment[SSLPlaintext.length]
+        private static string GetSniFromSslPlainText(ReadOnlySpan<byte> sslPlainText)
+        {
+            // Is SSL 3 handshake? SSL 2 does not support extensions - skipping as well
+            if (sslPlainText.Length < 5 || sslPlainText[0] != 0x16)
+            {
+                return null;
+            }
+
+            int handshakeLength = ReadUint16(sslPlainText.Slice(3));
+            ReadOnlySpan<byte> sslHandshake = sslPlainText.Slice(5);
+
+            if (handshakeLength != sslHandshake.Length)
+            {
+                return null;
+            }
+
+            return GetSniFromSslHandshake(sslHandshake);
+        }
+
+        // https://tools.ietf.org/html/rfc6101#section-5.6
+        // Handshake structure:
+        //   - HandshakeType msg_type (1 bytes) => 0x01 is client_hello
+        //   - uint24 length
+        //   - <msg_type> body
+        private static string GetSniFromSslHandshake(ReadOnlySpan<byte> sslHandshake)
+        {
+            // If not client hello then skip
+            if (sslHandshake.Length < 4 || sslHandshake[0] != 0x01)
+            {
+                return null;
+            }
+
+            int clientHelloLength = ReadUint24(sslHandshake.Slice(1));
+            ReadOnlySpan<byte> clientHello = sslHandshake.Slice(4);
+
+            if (clientHello.Length != clientHelloLength)
+            {
+                return null;
+            }
+
+            return GetSniFromClientHello(clientHello);
+        }
+
+        // 5.6.1.2. https://tools.ietf.org/html/rfc6101#section-5.6.1 - describes basic structure
+        // 2.1. https://www.ietf.org/rfc/rfc3546.txt - describes extended structure
```

Comment:
Do you mean https://tools.ietf.org/html/rfc3546#section-2.1 for this line?  The section number before the URI is a bit weird.